Services/ReplayFileService.cs

```diff
@@ -0,0 +1,174 @@
+            }
+
+            IDataRecord item;
+            try
+            {
+                item = await this.replayFilesStorage.GetAsync(id);
+            }
+            catch (ResourceNotFoundException)
+            {
+                throw;
+            }
+            catch (Exception e)
+            {
+                this.log.Error("Unable to load replay file from storage", () => new { id, e });
+                throw new ExternalDependencyException("Unable to load device model script from storage", e);
+            }
+
+            try
+            {
+                var deviceModelScript = JsonConvert.DeserializeObject<DataFile>(item.GetData());
+                deviceModelScript.ETag = item.GetETag();
+                return deviceModelScript;
+            }
+            catch (Exception e)
+            {
+                this.log.Error("Unable to parse device model script loaded from storage", () => new { id, e });
+                throw new ExternalDependencyException("Unable to parse device model script loaded from storage", e);
+            }
+        }
+
+        /// <summary>
+        /// Create a device model script.
+        /// </summary>
+        public async Task<DataFile> InsertAsync(DataFile replayFile)
+        {
+            replayFile.Created = DateTimeOffset.UtcNow;
+            replayFile.Modified = replayFile.Created;
+
+            if (string.IsNullOrEmpty(replayFile.Id))
+            {
+                replayFile.Id = Guid.NewGuid().ToString();
+            }
+
+            this.log.Debug("Creating a new replay file.", () => new { replayFile });
+
+            try
+            {
+                IDataRecord record = this.replayFilesStorage.BuildRecord(replayFile.Id,
+                    JsonConvert.SerializeObject(replayFile));
+
+                var result = await this.replayFilesStorage.CreateAsync(record);
+
+                replayFile.ETag = result.GetETag();
+            }
+            catch (Exception e)
+            {
+                this.log.Error("Failed to insert new replay file into storage",
+                    () => new { replayFile, e });
+                throw new ExternalDependencyException(
+                    "Failed to insert new replay file into storage", e);
+            }
+
+            return replayFile;
+        }
+
+        /// <summary>
+        /// Validate replay file
+        /// </summary>
+        public string Validate(Stream stream)
+        {
+            var reader = new StreamReader(stream);
+            var file = reader.ReadToEnd();
+
+            using (TextFieldParser parser = new TextFieldParser(file))
+            {
+                parser.TextFieldType = FieldType.Delimited;
+                parser.SetDelimiters(",");
+                while (!parser.EndOfData)
+                {
+                    try
+                    {
+                        string[] lines = parser.ReadFields();
```

Comment:
Why are we reading each line after we've already called reader.ReadToEnd()? Does the file object change at all between when we call ReadToEnd and when we return it?

Reply:
This will read the file and the next block validates each line is in valid format.